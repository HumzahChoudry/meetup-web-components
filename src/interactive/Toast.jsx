import React from 'react';
import cx from 'classnames';

import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';
import Icon from '../media/Icon';

/**
 * @module Toast
 */
class Toast extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			showToast: true
		};

		this._handleDismiss = this._handleDismiss.bind(this);
	}

	_handleDismiss() {
		if (this.props.setDismissedToast) {
			this.props.setDismissedToast(this);
		}

		this.setState({
			showToast: false
		});
	}

	render() {
		const {
			children,
			className,
			message,
			action,
			actionLabel,
			dismissable,
			setDismissedToast, // eslint-disable-line no-unused-vars
			autodismiss, // eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'toast padding--all',
			{
				'toast--hide': !this.state.showToast
			},
			className
		);

		return (
			<span
				className={classNames}
				{...other}>
					<Flex>
						<FlexItem>{message}</FlexItem>
						{action && (
							<FlexItem shrink>
								<a href='#' onClick={action} className='link'>
									{actionLabel}
								</a>
							</FlexItem>
						)}
						{dismissable && (
							<FlexItem
								shrink
								onClick={this._handleDismiss}
							>
								<Icon shape='cross' size='s' />
							</FlexItem>
						)}

					</Flex>
					{children}
			</span>
		);
	}
}

Toast.propTypes = {
	message: React.PropTypes.oneOfType([
		React.PropTypes.element,
		React.PropTypes.string
	]).isRequired,
	action: React.PropTypes.func,
	actionLabel: React.PropTypes.oneOfType([
		React.PropTypes.element,
		React.PropTypes.string
	]),
	dismissable: React.PropTypes.bool,
	autodismiss: React.PropTypes.bool
};

Toast.defaultProps = {
	autodismiss: true
};

export default Toast;
