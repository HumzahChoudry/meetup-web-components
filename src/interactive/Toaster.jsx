import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

/**
 * @module Toaster
 */
class Toaster extends React.Component {
	constructor(props) {
		super(props);

		this.timeouts = [];
		this.dismissedToasts = [];

		this.cloneToast = this.cloneToast.bind(this);
		this.setDismissedToast = this.setDismissedToast.bind(this);
		this.clearTimeouts = this.clearTimeouts.bind(this);
		this.mouseEnter = this.mouseEnter.bind(this);
		this.mouseLeave = this.mouseLeave.bind(this);

		this.state = {
			toasts: this.props.toasts.map(this.cloneToast),
			isHovered: false
		};
	}

	setDismissedToast(dismissedToast) {
		this.setState({
			toasts: this.state.toasts.filter(toast => dismissedToast.props.id !== toast.props.id)
		});
	}

	mouseEnter() {
		this.timeouts.forEach(clearTimeout);
	}

	mouseLeave() {
		this.setTimer();
	}

	setTimer() {
		const toastsToDismiss = this.state.toasts.filter((toast) => toast.props.autodismiss);

		toastsToDismiss.forEach((toast, i) => {
			this.timeouts.push(setTimeout(() => {
				this.setDismissedToast(toast);
			}, 1000*(i+1)));
		});
	}
	// end

	clearTimeouts() {
		this.timeouts.forEach(clearTimeout);
	}

	componentWillUnmount() {
		this.clearTimeouts();
	}

	componentDidMount() {
		this.setTimer();
	}

	/**
	 * @param {Object} Toast - `Toast` components to clone
	 * @param {number} i - index of the `toast`
	 * @returns {Array} `Toast` components with props from `Toaster`
	 */
	cloneToast(toast, i) {
		const toastProps = {
			key: i,
			id: toast.props.id || i,
			className: toast.props.className,
			message: toast.props.message,
			action: toast.props.action,
			actionLabel: toast.props.actionLabel,
			dismissable: toast.props.dismissable,
			autodismiss: toast.props.autodismiss,
			setDismissedToast: this.setDismissedToast
		};
		return React.cloneElement(toast, toastProps);
	}

	/**
	 * @returns {Array} `Toast` components
	 */
	renderToasts() {
		this.state.toasts = this.state.toasts.map(this.cloneToast);
		return this.state.toasts;
	}

	render() {
		const {
			className,
			toasts, // eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'toaster',
			{
				['hoveringggggg']: this.state.isHovered
			},
			className
		);

		return (
			<div
				className={classNames}
				onMouseEnter={this.mouseEnter}
				onMouseLeave={this.mouseLeave}
				{...other}>
				{
					this.renderToasts().map((toast, i) => (
						toast
					))
				}
			</div>
		);
	}
}

Toaster.propTypes = {
	toasts: PropTypes.arrayOf(PropTypes.element).isRequired
};

export default Toaster;
