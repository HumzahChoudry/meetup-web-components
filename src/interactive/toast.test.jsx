import React from 'react';
import { shallow, mount } from 'enzyme';
import Toast, {
	TOAST_CLASS,
	TOAST_ACTION_CLASS,
	TOAST_DISMISS_BTN_CLASS,
	SUCCESS_TOAST_CLASS,
	ERROR_TOAST_CLASS
} from './Toast';
import Toaster from './Toaster';

const ToastWithProps = (passedProps) => (
	<Toaster
		toasts={[
			<Toast {...passedProps}>
				Your toast is ready
			</Toast>
		]}
	/>
);

const ToastWithAction = ({action, actionLabel}) => (
	<Toaster
		toasts={[
			<Toast
				action={action}
				actionLabel={actionLabel}
				autodismiss={false}
			>
				Your toast is ready
			</Toast>
		]}
	/>
);

describe('Toaster', function() {
	const testToast = (
		<Toast>
			Your toast is ready
		</Toast>
	);

	const component = shallow(
		<Toaster
			toasts={[
				testToast
			]}
		/>
	);

	it('should store toasts in state', function(){
		expect(component.state().toasts.length).toBe(1);
	});

	it('should handle mouseEnter and mouseLeave', function(){
		const mouseEnterSpy = spyOn(Toaster.prototype, 'mouseEnter');
		const mouseLeaveSpy = spyOn(Toaster.prototype, 'mouseLeave');
		const toasterComponent = shallow(
			<Toaster
				toasts={[
					<Toast>
						Your toast is ready
					</Toast>
				]}
			/>
		);

		toasterComponent.simulate('mouseEnter');
		expect(mouseEnterSpy).toHaveBeenCalled();
		toasterComponent.simulate('mouseLeave');
		expect(mouseLeaveSpy).toHaveBeenCalled();
	});

	it('should set dismiss toasts', function(){
		component.instance().setDismissedToast(testToast);

		expect(component.state().toasts).not.toContain(testToast);
	});

	it('should dismiss toasts after specified time', function(){
		expect(component.state().toasts.length).toBe(1);
		setTimeout(() => {
			expect(component.state().toasts.length).toBe(0);
		}, 3000);
	});

	it('should set a timeout when the component mounts', function(){
		// mount component
		// check that length of `this.timeouts` is > 0
	});

	it('should clear the timeouts when specified', function(){
		// mount component
		// call `this.clearTimeouts()`
		// check that length of `this.timeouts` is 0
	});

});

describe('Toast', function() {
	const testToast = (
		<Toast>
			Your toast is ready
		</Toast>
	);

	const component = shallow(
		<Toaster
			toasts={[
				testToast
			]}
		/>
	);

	it('renders into the DOM', () => {
		expect(component).toMatchSnapshot();
	});

	it(`has a class of ${SUCCESS_TOAST_CLASS} when success prop is set`, () => {
		const component = mount(<ToastWithProps success />);
		const toast = component.find(`.${TOAST_CLASS}`);

		expect(toast.hasClass(SUCCESS_TOAST_CLASS)).toBe(true);
	});

	it(`has a class of ${ERROR_TOAST_CLASS} when success prop is set`, () => {
		const component = mount(<ToastWithProps error />);
		const toast = component.find(`.${TOAST_CLASS}`);

		expect(toast.hasClass(ERROR_TOAST_CLASS)).toBe(true);
	});

	it('should render an action if one is passed in', () => {
		const action = jest.fn();
		const component = mount(<ToastWithAction action={action} actionLabel='Action' />);
		const actionBtn = component.find(`.${TOAST_ACTION_CLASS}`);

		expect(actionBtn.length).toBe(1);
	});

	it('should call the function in the action when clicked', () => {
		const action = jest.fn();
		const component = mount(<ToastWithAction action={action} actionLabel='Action' />);
		const actionBtn = component.find(`.${TOAST_ACTION_CLASS}`);

		expect(action).not.toHaveBeenCalled();
		actionBtn.simulate('click');
		expect(action).toHaveBeenCalled();
	});

	it('should remove the Toast from the Toaster `toasts` state when the dismiss button is clicked', () => {
		// gotta figure this one out

		// const mountedComponent = mount(<ToastWithProps />);
		// const dismissBtn = mountedComponent.find(`.${TOAST_DISMISS_BTN_CLASS}`);

		// console.log(component.state('toasts').length);

		// expect(component.instance().state.toasts.length).toBe(1);
		// dismissBtn.simulate('click');
		// expect(component.instance().state.toasts.length).toBe(0);
	});

	it('should not render a dismiss button if dismissable prop is false', () => {
		const component = mount(<ToastWithProps dismissable={false} />);
		const dismissBtn = component.find(`.${TOAST_DISMISS_BTN_CLASS}`);

		expect(dismissBtn.length).toBe(0);
	});

});
