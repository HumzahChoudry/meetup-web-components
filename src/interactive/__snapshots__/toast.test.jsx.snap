// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders into the DOM 1`] = `
<div
  aria-live="assertive"
  className="toaster"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <CSSTransition
      appear={true}
      classNames="slide"
      key="csstrns-0"
      timeout={250}
    >
      <Toast
        autodismiss={true}
        dismissToast={[Function]}
        dismissable={true}
        id="toast-2"
        key="0"
      >
        Your toast is ready
      </Toast>
    </CSSTransition>
  </TransitionGroup>
</div>
`;
